### PR TITLE
Support detecting parameters in queries

### LINKED ISSUE #39: Parameter detection?
I have a terrible awful regex system for detecting params right now. It feels like it wouldn't be super hard to do in this library.

@MasterOdin  What do you think? Possible candidate for a v3?

### PR DESCRIPTION
**Pull request by rathboma**
This wasn't actually very hard honestly.

Fix #39 

Limitations:
- doesn't know if the parameters are in the 'right' place. Eg you can't use a parameter in place of an identifier, but it doesn't know that. Eg, this is not valid SQL: `select ? from foo`, but this is: `select x from foo where y = ?`

@MasterOdin let me know what you think. This doesn't actually break the v2 API, just adds a new field.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,5 @@
 package-lock.json
-yarn.lock
```
Comment (MasterOdin) [inline .gitignore:2]: I am of the opinion to ignore lock files for library packages as bumping lockfile versions isn't super useful. We also probably want to (ideally) set up the CI to test on the lowest support versions of packages as well as highest, which a lockfile would just get in the way of.

File: src/defines.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ export interface IdentifyResult {
   text: string;
   type: StatementType;
   executionType: ExecutionType;
+  parameters?: string[];
```
Comment (MasterOdin) [inline src/defines.ts:40]: I think I'd prefer just making this a permanent field to the result. Easier to deal with on the consumer code, and I don't consider adding an additional field to the return result a breaking change.

File: src/parser.ts [hunk 3]
```diff
@@ -570,6 +572,11 @@ function stateMachineStatementParser (statement: Statement, steps: Step[], { isS
         return;
       }
 
+      if (token.type === 'parameter') {
+        statement.parameters ||= []
```
Comment (MasterOdin) [inline src/parser.ts:576]: Just want to confirm that typescript compiles this down to something that isn't logical or as that's not supported on Node < 15, though again, totally fine (and would prefer) just making it part of `INITIAL_STATEMENT` and always be there for simplicity.

File: test/tokenizer/index.spec.ts [hunk 4]
```diff
@@ -263,4 +263,16 @@ describe('scan', function () {
...
+    const input = "?"
+    const actual = scanToken(initState(input))
+    const expected = {
+      type: 'parameter',
+      value: '?',
+      start: 0,
+      end: 0
+    }
+    expect(actual).to.eql(expected)
+  })
```
Comment (MasterOdin) [inline test/tokenizer/index.spec.ts:277]: Can you add more tests around the different dialects here?

**Review by MasterOdin — CHANGES_REQUESTED** [review]

**Comment by MasterOdin** [comment]
Yeah, the limitation is centered around the fact that we aren't implementing a true AST parser here. I have plans to work on a general (with dialect support) SQL AST parser in JS, just haven't found the time yet to do it, which would certainly be a v3 or v4 addition to be more precise in parsing.

**Review by MasterOdin — APPROVED** [review]

**Comment by rathboma** [comment]
Thanks for doing the extra work on this! Been a crazy few weeks.

**Pull request merged**